### PR TITLE
Delete feature for comments

### PR DESCRIPTION
**Pull request by anamariamilcu**
Added buttons for deletion for each comment, made button stylish to look like a generic delete image

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DeleteCommentServlet.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+public class DeleteCommentServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    long id = Long.parseLong(request.getParameter("id"));
+
+    Key commentEntityKey = KeyFactory.createKey("Comment", id);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.delete(commentEntityKey);
+  }
```
Comment (anamariamilcu) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteCommentServlet.java:38]: I tried to make his function to be 'doDelete' for the DELETE request

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -48,3 +48,47 @@ function addPhoto() {
...
+  liElement.appendChild(deleteButtonElement);
+  liElement.appendChild(usernameElement);
+  liElement.appendChild(commentElement);
+  return liElement;
+}
+
+function deleteComment(comment) {
+  const params = new URLSearchParams();
+  params.append('id', comment.id);
+  fetch('/delete-comment', {method: 'POST', body: params});
```
Comment (anamariamilcu) [inline portfolio/src/main/webapp/script.js]: Also here, I tried to set the method to be DELETE, but I think the mistake was related to how I set the body part.

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package com.google.sps.data;
+
+/* Class that represents a comment object */
+
+public class Comment {
+  private final long id;
+  private final String username;
+  private String text;
```
Comment (alexcepoi) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: Unclear why this is not final, yet id and username are. Now that you deleted setText, you can make this final I think.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -30,7 +30,6 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
```
Comment (alexcepoi) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:33]: Why did you remove this comment explaining what the servelet does?

Methods and classes should generally have documentation.

File: portfolio/src/main/webapp/style.css [hunk 5]
```diff
@@ -65,14 +65,42 @@ ul {
...
+.delete-button {
+  margin-top: 10px;
+  width: 26px;
+  grid-area: right;
+  align-content: center;
+  cursor: pointer;
+}
+
+.username-container {
+  grid-area: header;
```
Comment (alexcepoi) [inline portfolio/src/main/webapp/style.css:105]: As discussed, you are bundling into a PR which is meant to implement delete functionality, some unrelated changes to do with refactoring styling from js to css.

This is likely because you reused the same branch across PRs. While git supports pretty much anything, the typical development workflow is to work with multiple 'feature branches' which track the 'master' branch. Each feature branch ends up in a PR, and gets deleted after the PR gets merged.

This has several advantages:
- Assume we decide moving the styling from js to css was a big mistake. Untangling that from the current branch which also implements a new feature is now loads of work. With 2 separate branches, it's as easy as a git delete branch command.
- PR are super simple and focused. Hence they're super easy and quick to review and understand as each line changed is motivated by the PR description. Right now it's unclear if a change is because of the feature you're implementing or because of some unrelated reason related to some refactoring mentioned in a different PR :)

The downside is merge conflicts are more likely. Sadly merge conflicts are a part of life and it is possible to get good at dealing with them. There's also a more advanced option of stacking PR on top of each other, so that they still get reviewed individually, but reduces the chances of merge conflicts.

Don't worry about this for this PR.

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -63,18 +63,32 @@ function getCommentSectionFromServer() {
...
   // This container is used for showing the text of the comment.
-  const commentElement = document.createElement('p');
+  const commentElement = document.createElement('div');
+  commentElement.setAttribute('class', 'comment-text');
+  // Made an image represent the delete button.
+  const deleteButtonElement = document.createElement('img');
+  deleteButtonElement.setAttribute('class', 'delete-button');
+  deleteButtonElement.src = 'images/trash.png';
+  deleteButtonElement.addEventListener('click', () => {
+    console.log('here');
```
Comment (alexcepoi) [inline portfolio/src/main/webapp/script.js]: I think these were meant as debug logs and you intended to remove them before sending as a PR? :)

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -63,18 +63,32 @@ function getCommentSectionFromServer() {
...
+    deleteComment(comment);
+    // Remove the comment from the DOM.
+    liElement.remove();
```
Comment (alexcepoi) [inline portfolio/src/main/webapp/script.js]: I think this will delete the comment from the DOM even if the request fails. Ideally you would actually remove it from the DOM after the request returns successfully (via the callback).

Try fixing this in this PR.

**Review by alexcepoi — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (anamariamilcu) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:33]: Because it was part of the example in the tutorial, I will add a new updated one :)

File: portfolio/src/main/webapp/script.js (hunk 6 above)
Comment (anamariamilcu) [reply portfolio/src/main/webapp/script.js]: yes, oops

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -63,18 +63,35 @@ function getCommentSectionFromServer() {
...
-  commentElement.style.height = '100%';
+  liElement.appendChild(deleteButtonElement);
   liElement.appendChild(usernameElement);
   liElement.appendChild(commentElement);
-  liElement.style.border = 'dotted pink';
   return liElement;
 }
+
+function deleteComment(comment, element) {
+  let resultStatus;
```
Comment (alexcepoi) [inline portfolio/src/main/webapp/script.js]: unused?

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -63,18 +63,35 @@ function getCommentSectionFromServer() {
...
-  commentElement.style.fontWeight = 'initial';
-  commentElement.style.height = '100%';
+  liElement.appendChild(deleteButtonElement);
   liElement.appendChild(usernameElement);
   liElement.appendChild(commentElement);
-  liElement.style.border = 'dotted pink';
   return liElement;
 }
+
+function deleteComment(comment, element) {
```
Comment (alexcepoi) [inline portfolio/src/main/webapp/script.js:89]: What exactly this functoin does is not entirely obvious.

Consider an online comment. Most functions should have an oneline comment.

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -63,18 +63,35 @@ function getCommentSectionFromServer() {
...
+  let resultStatus;
+  const params = new URLSearchParams();
+  params.append('id', comment.id);
+  fetch('/delete-comment', {method: 'POST', body: params}).then(result => {
+    //Check if the comment was deleted succesfully.
+    if (result.status === 200) {
```
Comment (alexcepoi) [inline portfolio/src/main/webapp/script.js:93]: Potentially add a TODO to handle errors correctly.

You don't need to fix TODOs immediately (and adding a TODO does not even bind you from ever fixing it) but it signals to a future reader (who may be wondering why is there no error handling here), that yes, the error handling was not implemented (yet).

**Review by alexcepoi — APPROVED** [review]

File: portfolio/src/main/webapp/script.js (hunk 10 above)
Comment (anamariamilcu) [reply portfolio/src/main/webapp/script.js:93]: But in this case, what should be done to handle the possible error? The comment should not be deleted in that case

Comment (alexcepoi) [reply portfolio/src/main/webapp/script.js:93]: I meant, you probably want to inform the user somehow that there was an error.
Yes, the comments should not be deleted :)
